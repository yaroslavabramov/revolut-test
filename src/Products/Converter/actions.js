import { SUBSCRIBE_DATA, UPDATE_RATES, CANCEL_SUBSCRIBE } from './constants';

const subscribeData = () => ({
  type: SUBSCRIBE_DATA
});

const cancelSubscribe = () => ({
  type: CANCEL_SUBSCRIBE
});

const updateRates = rates => ({
  type: UPDATE_RATES,
  rates
});

export { subscribeData, cancelSubscribe, updateRates };
