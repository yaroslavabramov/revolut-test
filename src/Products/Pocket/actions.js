import { UPDATE_FIELD, EXCHANGE_VALUES } from './constants';

const updateField = (field, value) => ({
  type: UPDATE_FIELD,
  value,
  field
});

const exchangeValues = (from, to) => ({
  type: EXCHANGE_VALUES,
  from,
  to
});

export { updateField, exchangeValues };
