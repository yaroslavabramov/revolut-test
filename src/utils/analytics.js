import Analytics from 'analytics';
import googleAnalyticsPlugin from '@analytics/google-analytics';

export const analytics = Analytics({
  app: 'my-app-name',
  version: 1,
  plugins: [
    googleAnalyticsPlugin({
      trackingId: 'UA-164609686-1'
    })
  ]
});
