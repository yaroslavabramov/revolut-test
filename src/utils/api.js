const getRates = async () => {
  try {
    const response = await fetch(
      'https://openexchangerates.org/api/latest.json?app_id=bb16478b4da442c99aafe75fd3d13158'
    );
    if (response.ok) return await response.json();
    else throw new Error('response is not ok');
  } catch (e) {
    throw new Error(e);
  }
};

export { getRates };
