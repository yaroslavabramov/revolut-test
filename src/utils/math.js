export const round = number => Math.round(number * 100) / 100;
