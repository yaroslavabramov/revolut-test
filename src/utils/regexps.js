export const numbericRegexp = /^\d*\.?\d?\d?$/;
