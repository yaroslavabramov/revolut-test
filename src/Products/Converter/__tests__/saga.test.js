import { call, delay, put, takeEvery, race, take } from 'redux-saga/effects';
import {
  SUBSCRIBE_RATES,
  CANCEL_SUBSCRIPTION,
  UPDATE_FIELD_CURRENCY,
  UPDATE_VALUE_FROM_INPUT,
  UPDATE_RATES
} from '../constants';
import converterSaga, {
  subscribeRates,
  recountValues,
  debouncedRecountValues,
  startSubscription,
  fetchRates
} from '../saga';
import { getRates } from '../../../utils/api';
import { updateRates, updateFieldValue } from '../actions';

describe('converter saga', () => {
  describe('default saga. should set listeners', () => {
    const generator = converterSaga();
    it('should set listener for SUBSCRIBE_RATES aciton', () => {
      const listenerDescriptor = generator.next().value;
      expect(listenerDescriptor).toEqual(
        takeEvery(SUBSCRIBE_RATES, subscribeRates)
      );
    });
    it('should set listener for UPDATE_FIELD_CURRENCY and UPDATE_RATES action', () => {
      const listenerDescriptor = generator.next().value;
      expect(listenerDescriptor).toEqual(
        takeEvery([UPDATE_FIELD_CURRENCY, UPDATE_RATES], recountValues)
      );
    });
    it('should set listener for UPDATE_VALUE_FROM_INPUT aciton', () => {
      const listenerDescriptor = generator.next().value;
      expect(listenerDescriptor).toEqual(
        takeEvery(UPDATE_VALUE_FROM_INPUT, debouncedRecountValues)
      );
    });
  });
  describe('subscribeRates saga', () => {
    const generator = subscribeRates();
    it('should start race', () => {
      const raceDescriptor = generator.next().value;
      expect(raceDescriptor).toEqual(
        race({
          subscribe: call(startSubscription),
          cancel: take(CANCEL_SUBSCRIPTION)
        })
      );
    });
  });
  describe('fetchRates saga', () => {
    let generator;
    beforeEach(() => {
      generator = fetchRates();
    });
    it('should fetch rates', () => {
      const requestDescriptor = generator.next().value;
      expect(requestDescriptor).toEqual(getRates());
    });
    it('should send data to store', () => {
      generator.next();
      const rates = {
        EUR: '123',
        GBP: '234'
      };
      const putDescriptor = generator.next({ rates }).value;
      expect(putDescriptor).toEqual(put(updateRates(rates)));
    });
    it('should log error in conole', () => {
      generator.next();
      const error = new Error('smth wrong!');
      const logDescriptor = generator.throw(error).value;
      expect(logDescriptor).toEqual(call(console.error, error));
    });
  });
  describe('recount values saga', () => {
    let generator;
    beforeEach(() => {
      generator = recountValues();
    });
    it('should select active field', () => {
      const selectorDescriptor = generator.next().value;
      expect(selectorDescriptor).toMatchSnapshot();
    });
    it('should select coefficient', () => {
      generator.next();
      const selectorDescriptor = generator.next('from').value;
      expect(selectorDescriptor).toMatchSnapshot();
    });
    it('should select from value', () => {
      generator.next();
      generator.next('from');
      const selectorDescriptor = generator.next(0.4).value;
      expect(selectorDescriptor).toMatchSnapshot();
    });
    it('should select to value', () => {
      generator.next();
      generator.next('to');
      const selectorDescriptor = generator.next(0.4).value;
      expect(selectorDescriptor).toMatchSnapshot();
    });
    it('should put new data to store for from field', () => {
      generator.next();
      generator.next('from');
      generator.next(0.4);
      const putDescriptor = generator.next(123).value;
      const result = Math.round(123 * 0.4 * 100) / 100;
      expect(putDescriptor).toEqual(
        put(updateFieldValue('to', result.toString()))
      );
    });
    it('should put new data to store for to field', () => {
      generator.next();
      generator.next('to');
      generator.next(0.4);
      const putDescriptor = generator.next(123).value;
      const result = Math.round((123 / 0.4) * 100) / 100;
      expect(putDescriptor).toEqual(
        put(updateFieldValue('from', result.toString()))
      );
    });
  });
  describe('debouncedRecountValues saga', () => {
    const generator = debouncedRecountValues();
    it('should delay for 100ms', () => {
      const delayDescriptor = generator.next().value;
      expect(delayDescriptor).toEqual(delay(100));
    });
    it('should call recountValues', () => {
      const delayDescriptor = generator.next().value;
      expect(delayDescriptor).toEqual(call(recountValues));
    });
  });
});
