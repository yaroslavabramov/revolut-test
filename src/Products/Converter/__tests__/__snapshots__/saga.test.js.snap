// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converter saga recount values saga should select active field 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`converter saga recount values saga should select coefficient 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`converter saga recount values saga should select from value 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`converter saga recount values saga should select to value 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
